File: src/main/kotlin/io/titandata/titan/providers/Metadata.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers

import io.titandata.titan.clients.Docker.Companion.toList

data class Image(val image: String, val tag: String?, val digest: String)
data class Port(val protocol: String, val port: String)
data class Volume(val name: String, val path: String)

data class Metadata(
    val version: Version,
    var user: String?,
    var email: String?,
    var message: String?,
    var source: String?,
    var tags: Map<String, String>?,
    var timestamp: String?,
    val image: Image,
    val environment: List<String>,
    val ports: List<Port>,
    val volumes: List<Volume>
) {
    enum class Version(val value: String) {
        V2("v2"),
        V1("v1")
    }
    companion object{
        fun Metadata.toMap(): Map<String, Any> {
            val returnMap = mutableMapOf<String, Any>()
            if (!this.user.isNullOrEmpty()) {
                returnMap["user"] = this.user!!
            }
            if (!this.email.isNullOrEmpty()) {
                returnMap["email"] = this.email!!
            }
            if (!this.message.isNullOrEmpty()) {
                returnMap["message"] = this.message!!
            }
            if (!this.source.isNullOrEmpty()) {
                returnMap["source"] = this.source!!
            }
            if (!this.tags.isNullOrEmpty()) {
                returnMap["tags"] = this.tags!!
            }
            if (!this.timestamp.isNullOrEmpty()) {
                returnMap["timestamp"] = this.timestamp!!
            }
            if (this.version == Version.V2) {
                returnMap["v2"] = mapOf(
                    "image" to this.image,
                    "environment" to this.environment,
                    "ports" to this.ports,
                    "volumes" to this.volumes
                )
            }
            if (this.version == Version.V1) {
                returnMap["container"] = this.image.digest
                returnMap["image"] = this.image.image
                returnMap["tag"] = this.image.tag!!
                returnMap["digest"] = this.image.digest
            }
            return returnMap
        }
        fun load(metaMap: Map<String, Any>): Metadata {
            return when {
                metaMap.containsKey("v2") -> mapV2(metaMap)
                else -> mapV1(metaMap)
            }
        }
        private fun mapV2(metaMap: Map<String, Any>): Metadata {
            val user = when {
                metaMap.containsKey("user") -> metaMap["user"] as String
                else -> null
            }
            val email = when {
                metaMap.containsKey("email") -> metaMap["email"] as String
                else -> null
            }
            val message = when {
                metaMap.containsKey("message") -> metaMap["message"] as String
                else -> null
            }
            val source = when {
                metaMap.containsKey("source") -> metaMap["source"] as String
                else -> null
            }
            @Suppress("UNCHECKED_CAST")
            val tags = when {
                metaMap.containsKey("tags") -> metaMap["tags"] as Map<String, String>
                else -> null
            }
            val timestamp = when {
                metaMap.containsKey("timestamp") -> metaMap["timestamp"] as String
                else -> null
            }

            @Suppress("UNCHECKED_CAST")
            val meta = metaMap["v2"] as Map<String, Any>

            @Suppress("UNCHECKED_CAST")
            val imageMap = meta["image"] as Map<String, String>
            val image = Image(imageMap["image"]!!, imageMap["tag"]!!, imageMap["digest"]!!)

            @Suppress("UNCHECKED_CAST")
            val environment =  meta["environment"] as List<String>

            @Suppress("UNCHECKED_CAST")
            val mapPorts = meta["ports"] as List<Map<String, String>>
            val ports = mutableListOf<Port>()
            for (mapPort in mapPorts) {
                ports.add(
                        Port(mapPort["protocol"]!!, mapPort["port"]!!)

                )
            }
            @Suppress("UNCHECKED_CAST")
            val mapVols = meta["volumes"] as List<Map<String, String>>
            val volumes = mutableListOf<Volume>()
            for (mapVol in mapVols) {
                volumes.add(Volume(mapVol["name"]!!, mapVol["path"]!!))
            }
            return Metadata(Version.V2, user, email, message, source, tags, timestamp, image, environment, ports, volumes)
        }

        private fun mapV1(metaMap: Map<String, Any>): Metadata {
            val user = when {
                metaMap.containsKey("user") -> metaMap["user"] as String
                else -> null
            }
            val email = when {
                metaMap.containsKey("email") -> metaMap["email"] as String
                else -> null
            }
            val message = when {
                metaMap.containsKey("message") -> metaMap["message"] as String
                else -> null
            }
            val source = when {
                metaMap.containsKey("source") -> metaMap["source"] as String
                else -> null
            }
            val tags = when {
                metaMap.containsKey("tags") -> metaMap["tags"] as Map<String, String>
                else -> null
            }
            val timestamp = when {
                metaMap.containsKey("timestamp") -> metaMap["timestamp"] as String
                else -> null
            }
            val digest = metaMap["container"] as String
            val imageName = when {
                metaMap.containsKey("image") -> metaMap["image"] as String
                else -> digest.split("@")[0]
            }
            val imageTag = when {
                metaMap.containsKey("tag") -> metaMap["tag"] as String
                else -> null
            }
            val image =  Image(imageName, imageTag, digest)
            val runtimeString = metaMap["runtime"] as String
            val runtime = runtimeString.removePrefix("[").removeSuffix("]").toList(", ")

            val envs = mutableListOf<String>()
            val ports = mutableListOf<Port>()
            val volumes = mutableListOf<Volume>()

            for ((index, item) in runtime.withIndex()) {
                if (item == "--env" || item == "-e") {
                    envs.add(runtime[index + 1])
                }
                if (item == "-p") {
                    val port = when {
                        runtime[index + 1].contains(":") -> runtime[index + 1].split(":")[1]
                        else -> runtime[index + 1]
                    }
                    val addPort = Port(
                            "tcp",
                            port
                    )
                    ports.add(addPort)
                }
                if (item == "--mount") {
                    val vols = runtime[index + 1].split(",")
                    var name = ""
                    var path = ""
                    for (vol in vols){
                        val splitVol = vol.split("=")
                        when (splitVol[0]) {
                            "src" -> name = splitVol[1].split("/")[1]
                            "dst" -> path = splitVol[1]
                            else -> "no match"
                        }
                    }
                    volumes.add(Volume(name, path))
                }
            }
            return Metadata(Version.V1, user, email, message, source, tags, timestamp, image, envs, ports, volumes)
        }
    }
}

File: internal/app/providers/common/Metadata.go
package common

import (
	"fmt"
	"strings"
)

type Version string

const(
	V2 Version = "v2"
	V1 Version = "v1"
)

type image struct {
	Image  string `json:"image"`
	Tag    string `json:"tag"`
	Digest string `json:"digest"`
}
type port struct {
	Protocol string `json:"protocol"`
	Port     string `json:"port"`
}
type volume struct {
	Name string `json:"name"`
	Path string `json:"path"`
}

type Metadata struct {
	version Version
	user string
	email string
	message string
	source string
	tags map[string]string
	timestamp string
	image image
	environment []interface{}
	ports []port
	volumes []volume
}

func (m *Metadata) SetUser(s string) {
	m.user = s
}

func (m *Metadata) SetEmail(s string)  {
	m.email = s
}

func (m *Metadata) SetMessage(s string) {
	m.message = s
}

func (m *Metadata) SetTags(t map[string]string) {
	m.tags = t
}

func (m *Metadata) SetSource(s string) {
	m.source = s
}

func (m Metadata) ToMap() map[string]interface{} {
	returnMap := make(map[string]interface{})
	if m.user != "" {
		returnMap["user"] = m.user
	}
	if m.email != "" {
		returnMap["email"] = m.email
	}
	if m.message != "" {
		returnMap["message"] = m.message
	}
	if m.source != "" {
		returnMap["source"] = m.source
	}
	if len(m.tags) > 0 {
		returnMap["tags"] = m.tags
	}
	if m.timestamp != "" {
		returnMap["timestamp"] = m.timestamp
	}
	if m.version == V2 {
		returnMap["v2"] = map[string]interface{} {
			"image": m.image,
			"environment": m.environment,
			"ports": m.ports,
			"volumes": m.volumes,
		}
	}
	if m.version == V1 {
		returnMap["container"] = m.image.Digest
		returnMap["image"] = m.image.Image
		returnMap["tag"] = m.image.Tag
		returnMap["digest"] = m.image.Digest
	}
	return returnMap
}

func (m Metadata) Load(metaMap map[string]interface{}) Metadata {
	_, ok := metaMap["v2"]
	if ok  {
		return m.MapV2(metaMap)
	} else {
		return m.MapV1(metaMap)
	}
}

func stringFromMap(m map[string]interface{}, k string) string {
	v, ok := m[k]
	if ok {
		return v.(string)
	} else {
		return ""
	}
}

func (m Metadata) MapV2(metaMap map[string]interface{}) Metadata {
	user := stringFromMap(metaMap, "user")
	email := stringFromMap(metaMap, "email")
	message := stringFromMap(metaMap, "message")
	source := stringFromMap(metaMap, "source")
	timestamp := stringFromMap(metaMap, "timestamp")

	tags := make(map[string]string)
	v, ok := metaMap["tags"].(map[string]interface{})
	if ok {
		for key, value := range v {
			switch value := value.(type) {
			case string:
				tags[key] = value
			}
		}
	} else {
		tags = nil
	}

	meta := metaMap["v2"].(map[string]interface{})
	imageMap := meta["image"].(map[string]interface{})
	image := image{
		Image:  fmt.Sprintf("%v", imageMap["image"]),
		Tag:    fmt.Sprintf("%v", imageMap["tag"]),
		Digest: fmt.Sprintf("%v", imageMap["digest"]),
	}

	envCheck := meta["environment"] //TODO this can be empty
	var environment []interface{}
	if envCheck != nil {
		environment = meta["environment"].([]interface{})
	} else {
		environment = nil
	}

	metaPorts := meta["ports"].([]interface{})
	var ports []port
	for _, v := range metaPorts{
		mapPort := v.(map[string]interface{})
		ports = append(ports, port{
			Protocol: fmt.Sprintf("%v", mapPort["protocol"]),
			Port:     fmt.Sprintf("%v", mapPort["port"]),
		})
	}
	var volumes []volume
	metaVols := meta["volumes"].([]interface{})
	for _, v := range metaVols {
		metaVol := v.(map[string]interface{})
		volumes = append(volumes, volume{
			Name: fmt.Sprintf("%v",metaVol["name"]),
			Path: fmt.Sprintf("%v",metaVol["path"]),
		})
	}

	return Metadata{
		version:     V2,
		user:        user,
		email:       email,
		message:     message,
		source:      source,
		tags:        tags,
		timestamp:   timestamp,
		image:       image,
		environment: environment,
		ports:       ports,
		volumes:     volumes,
	}
}

func (m Metadata) MapV1(metaMap map[string]interface{}) Metadata {
	user := stringFromMap(metaMap, "user")
	email := stringFromMap(metaMap, "email")
	message := stringFromMap(metaMap, "message")
	source := stringFromMap(metaMap, "source")
	timestamp := stringFromMap(metaMap, "timestamp")

	var tags map[string]string
	v, ok := metaMap["tags"]
	if ok {
		tags = v.(map[string]string)
	} else {
		tags = nil
	}

	digest := stringFromMap(metaMap, "container")
	var imageName string
	v, ok = metaMap["Image"]
	if ok {
		imageName = v.(string)
	} else {
		if strings.Contains(digest, "@") {
			imageName = strings.Split(digest, "@")[0]
		} else {
			imageName = strings.Split(digest, ":")[0]
		}
	}
	var imageTag string
	v, ok = metaMap["Tag"]
	if ok {
		imageTag = v.(string)
	} else {
		imageTag = ""
	}
	image := image{
		Image:  imageName,
		Tag:    imageTag,
		Digest: digest,
	}

	runtimeString := stringFromMap(metaMap, "runtime")
	runtimeString = strings.TrimLeft(runtimeString, "[")
	runtimeString = strings.TrimRight(runtimeString, "]")
	runtime := strings.Split(runtimeString, " ")
	for i, n := range runtime{
		if strings.Contains(n,"--mount") {
			runtime = append(runtime[:i], runtime[i+1:]...)
		}
		if strings.Contains(n,"type=volume"){
			runtime = append(runtime[:i], runtime[i+1:]...)
		}
	}

	var envs []interface{}
	var ports []port
	var volumes []volume

	for i, n := range runtime {
		if strings.Contains(n, "--env") || strings.Contains(n, "-e") {
			envs = append(envs, runtime[i+1])
		}
		if n == "-p" {
			var p string
			if strings.Contains(runtime[i+1], ":"){
				p = strings.Split(runtime[i+1], ":")[1]
			} else {
				p = runtime[i+1]
			}
			ports = append(ports, port{
				Protocol: "tcp",
				Port:     p,
			})
		}
		if n == "--mount" {
			vols := strings.Split(runtime[i+1], ",")
			var name string
			var path string
			for _, vol := range vols {
				sv := strings.Split(vol, "=")
				if sv[0] == "src" {
					name = strings.Split(sv[1], "/")[1]
				}
				if sv[0] == "dst" {
					path = sv[1]
				}
			}
			volumes = append(volumes, volume{
				Name: name,
				Path: path,
			})
		}
	}
	return Metadata{
		version:     V1,
		user:        user,
		email:       email,
		message:     message,
		source:      source,
		tags:        tags,
		timestamp:   timestamp,
		image:       image,
		environment: envs,
		ports:       ports,
		volumes:     volumes,
	}
}
